Sources/Basic/TemporaryFile.swift

```diff
@@ -86,31 +108,107 @@ public final class TemporaryFile {
 
     /// Remove the temporary file before deallocating.
     deinit { unlink(path) }
+}
+
+extension TemporaryFile: CustomStringConvertible {
+    public var description: String {
+        return "<TemporaryFile: \(path)>"
+    }
+}
+
+/// FIXME: This isn't right place to declare this, probably POSIX or merge with FSProxyError?
+/// Contains the error which can be thrown while creating a directory using POSIX's mkdir.
+public enum MakeDirectoryError: ErrorProtocol {
+    /// The given path already exists as a directory, file or symbolic link.
+    case pathExists
+    /// The path provided was too long.
+    case pathTooLong
+    /// Process does not have permissions to create directory.
+    /// Note: Includes read-only filesystems or if file system does not support directory creation.
+    case permissionDenied
+    /// The path provided is unresolvable because it has too many symbolic links or a path component is invalid.
+    case unresolvablePathComponent
+    /// Exceeded user quota or kernel is out of memory.
+    case outOfMemory
+    /// All other system errors with their value.
+    case other(Int32)
+}
+
+private extension MakeDirectoryError {
+    init(errno: Int32) {
+        switch errno {
+        case libc.EEXIST:
+            self = .pathExists
+        case libc.ENAMETOOLONG:
+            self = .pathTooLong
+        case libc.EACCES, libc.EFAULT, libc.EPERM, libc.EROFS:
+            self = .permissionDenied
+        case libc.ELOOP, libc.ENOENT, libc.ENOTDIR:
+            self = .unresolvablePathComponent
+        case libc.ENOMEM, libc.EDQUOT:
+            self = .outOfMemory
+        default:
+            self = .other(errno)
+        }
+    }
+}
+
+/// A class to create disposable directories using POSIX's mkdtemp() method.
+public final class TemporaryDirectory {
+    /// If specified during init, the temporary directory name begins with this prefix.
+    let prefix: String
+
+    /// The directory in which the temporary directory is created.
+    public let dir: String
```

Comment:
This is always going to be `path.parentDirectory`, right? If so I'm not sure it is worth having both, a simpler API seems nicer.